### DISCUSSION TITLE
Roll back the latest migration group

### DISCUSSION THREAD

**Discussion by vadim-shilov** [prisma/prisma#4617] [Q&A]
Will Prisma Migrate support to roll back the latest migration group?
I couldn't find anything like this http://knexjs.org/#Migrations-rollback in docs.

For example:
A new version of an app includes some migrations.
It was deployed (migrations were applied) but after was made a decision to roll it back to the previous version.
What the way to roll back the last batch of migrations?

**Comment by albertoperdomo** [comment]
Traditional migration systems rely on up and down migrations to move up and down version of the DB schema. While this system is prevalent, we've learned from speaking with many developers that most of the time down migrations are not used.

In a production environment, rollbacks can lead to data loss. For example: dropping a column or table that was added in a migration, and which has been exposed to users for storing data, will lead to losing this data from users. It may sound counterintuitive, but often the way to resolve this kind of situation is not by undoing the changes from the migrations but rather to fix the underlying issue, or disable the corresponding feature.

Our current recommendation is to resolve issues by rolling forward: create a new migration with the changes you want.

Note all of this is relevant when the migrations have been applied in other environments, i.e. production. If the migrations you want to _"undo"_ are in development, you can always delete the migration directory, and the next time you run `prisma migrate dev`, it will prompt to reset your development database.

Removing database migration directories is of course not something you should do in production or a similar environment, i.e. staging.

We are open to feedback about rollbacks (and of course about everything else): If you have strong opinions on why we should implement rollbacks please do share your thoughts with us.

**Reply by dukedorje** [reply]
What about in development when you don't want to fully reset?  It does take some effort to put in example data locally. It'd be really nice to be able to roll back, edit the schema, roll forward in development without fully deleting & re-creating all sample data.

**Reply by dara-rockwell** [reply]
This is a reasonable approach for production, but like @dukejones pointed out, I'd like finer control over rolling back in development.

**Reply by ryanpagel** [reply]
Yes, I've often had the same need in a dev environment.

**Reply by verekia** [reply]
Using a down migration does not necessarily result in data loss. If you quickly realize you need to rollback, it is much more comfortable and safe to simply run `migrate down` from the CLI than having a developer manually modify the Prisma schema, generate a new migration, and run it. With an auto-generated down migration, it is guaranteed that the down migration will undo the change exactly. With a fix migration you can mess it up and end up with a different wrong schema, particularly if the rollback is urgent and causes panic.

**Reply by dukedorje** [reply]
Yes!  Auto-generating is a can of worms though.  Ruby On Rails has a decent approach: Auto-generates for simple cases like create table, and has a format for manually specifying up and down for your own migration.  (With the option to have an IrreversibleMigration.) Helps development, and helps confidence in deploying to production!

**Reply by fabiohvp** [reply]
> What about in development when you don't want to fully reset? It does take some effort to put in example data locally. It'd be really nice to be able to roll back, edit the schema, roll forward in development without fully deleting & re-creating all sample data.

just create a new migration with the fix, if you want to keep your history clean then delete both of migrations and create a new one with everything fixed

**Reply by dara-rockwell** [reply]
@fabiohvp That would require a rollback of the previous migration. Rolling forward in a development environment is an unnecessary burden. I should be able to make a change, see how the migration went, then undo that change and try again as many times as it takes to get it right before pushing the code. I should be able to do this without wiping out my development data, which I may have spent a lot of time setting up.

**Reply by goldo** [reply]
I am currently working in a company with a lot of data per environment (stagings, preprod, etc.) and we cannot afford to reset our database every time we want to deploy a branch that hasnt rebase on another branch because of a migration it included.. 
If we had to make another migration to prevent the code from failing on 1 particular environment, it would mean more work, just a deploy a branch. I see this as a deal breaker, for us at least

**Reply by fny** [reply]
> just create a new migration with the fix, if you want to keep your history clean then delete both of migrations and create a new one with everything fixed

This sounds like me manually doing work instead of my computer doing it for me.  Every other migration system (i.e. TypeORM, sequelize, alembic, ActiveRecord) has this feature for a reason: It's incredibly useful and developer friendly. If Prisma aims to be a polished solution, this needs to be addressed.

We will be keeping TypeORM as our go to until this and a few other things get a bit more polished.

**Reply by RWOverdijk** [reply]
Is this still the recommendation? "Roll forward"? Or did you come to your senses 😄

**Reply by josephJaoNasser** [reply]
I have an issue wherein the I have a date field named ``` createdAt ``` which is a Date and I migrated it locally.

However, I forgot to put a ``` @default ``` value or set the ``` @default  ``` value to ``` now() ``` and I got an error when deploying to a staging DB saying that the migration cannot be done because ``` createdAt ``` had null values.

I thought this wasn't a big deal, I'll just "roll forward" then. So I updated my schema and added ```  @default(now()) ``` for my ``` createdAt ``` field.

However, as it was deploying, it got stuck on the previous migration wherein it tries to migrate but can't because ``` createdAt ``` has no default value. **I litterally can't migrate BOTH the fix and the faulty migration**  because the fix cannot be applied since it has to apply the faulty migration first!

Rolling backward should be an option in case we made a mistake and need to do one thing before the other. But if anybody knows how to do this, please reply.

My option right now is to manually remove the faulty migration folders, clear my local DB, uninstall/re-install Prisma, then create the migrations correctly this time and then deploy.

**Reply by JHIH-LEI** [reply]
sequelize support it !

**Reply by formigarafa** [reply]
> Traditional migration systems rely on up and down migrations to move up and down version of the DB schema. 
> ...

I bet those **traditional systems** rely on up and down based on arguments, not just because of opinions.
I can imagine two simple scenarios where being able to rollback a migration would be at least helpful:
- Developing multiple features in separated branches; and first and foremost
- Contingency plan: (expect for the unexpected).

And doesn't matter how much you test, prepare, check and review, one day mistakes will be made and whoever is using this tool they will hope they can rollback the migration as quick as possible to avoid build up of more of whatever problem they will be having form their production environment.

I don't think people are asking for autogenerated down migrations, which even the autogenerated up migration can be misleading. Some people will understand what I mean when they are adding a new non nullable column with a default value in a table with millions of records. Even though there is a single command to do that if you want this to happen with zero-downtime you will need to do this change in steps. Likely, in multiple transactions.

But maybe if there is, at least, a clear path on how to achieve this optionally, then this discussion wouldn't be necessary.
Something like a command that will execute the down migration to the latest migration executed if the file exists.
That would probably be enough.

**Reply by tylerslater-io** [reply]
> While this system is prevalent, we've learned from speaking with many developers that most of the time down migrations are not used.

Rollbacks are uncommon but are a requirement in any SDLC. Fixing forward leads to more issues and an increase in downtime. For quick prototyping and no users that's fine, but for a production system this is a dealbreaker to us implementing Prisma as our ORM. While many can argue that reverting a commit in git is less used than simply committing new code, it still plays an integral part in the use of the system.

>It may sound counterintuitive, but often the way to resolve this kind of situation is not by undoing the changes from the migrations but rather to fix the underlying issue, or disable the corresponding feature.

This definitely is counterintuitive as encouraging fixes against production leads to more downtime than a rollback where you can address the issue with more precision and intention.

Prisma seems to be growing in adoption and I would love to see it mature with the necessary tooling needed for a stable production system.

**Reply by schuhegger** [reply]
@albertoperdomo for us this feature is essential to be able to use Prisma not just for prototyping but also as stable production ORM. So far we love Prisma but unfortunately it would be a dealbreaker for us if we cannot roll back the database. Are there any news on this topic and is Prisma planning to integrate this feature soon?

**Reply by albertoperdomo** [reply]
@schuhegger Sorry, I no longer work at Prisma. cc @janpio

**Reply by janpio** [reply]
Here is our documentation on the current approach on down migrations and rollbacks: https://www.prisma.io/docs/guides/migrate/developing-with-prisma-migrate/generating-down-migrations

**Comment by orcioly** [comment]
Bom dia pessoal,

Verificando os comentários, seria bom se o Prisma V2 ou posterior quando surgir, ter a opção para reverter às última migrações. No modo de desenvolvimento se deletarmos o arquivo de migração e rodar o migrate dev resolveria o problema, porém, ficaríamos sem o histórico/versionamente dessa reversão.

Foi apenas meu ponto de vista e peço desculpas se eu estiver errado.

Grato,

**Reply by janpio** [reply]
Translation:

> Good morning people,
> 
> Checking the comments, it would be nice if Prisma V2 or later when it comes out has the option to revert to the last migrations. In development mode if we delete the migration file and run migrate dev it would solve the problem, however, we would be without the history/version of this rollback.
> 
> It was just my point of view and I apologize if I'm wrong.

**Comment by RohitRox** [comment]
Any new news on this one?

**Comment by orcioly** [comment]
Good night,

Not. It's complicated because I've been enjoying Prisma, but I realize that I'm getting limited with some things. The difficulty of clearly documenting libraries to work with Prisma is precarious and unclear. I'm honestly thinking of removing Prisma from the project and implementing TypeORM. I have no choice because I have a deadline to meet.

**Comment by eddienubes** [comment]
It's definitely something you should consider to implement!

Hope you didn't forget about this thread.

**Comment by gabimor** [comment]
run into a similar issue with a "bad" migration that already tried to run on staging. I couldn't add a new migration to solve it cause it never got there (failing on the original bad migration).
also, deleting the migration from `prisma/migrations` didn't do the trick as the old one still tried to run when deploying!
this is what i did:

- turns out that failed migrations leave a record in db`migration` table. I deleted that record. (just one record for the one bad migration. you will see the error stored there as well)
- delete the migration file and folder from the `prisma/migrations` folder
- generate a new migration with @default() (this will require you to reset the development db, meaning running all the migrations again).

after a commit and a push (which runs the CI/CD and the migrations), it worked.
the "bad" migration was gone and the new one run adding the field with a default value

@josephJaoNasser

**Reply by josephJaoNasser** [reply]
I've been doing this as well. The problem with this is that, as you said, it resets the development DB and it deletes the data that's been stored locally. 

Fortunately for me, I've been playing around with a staging DB deployed online as opposed to locally stored database files, so whenever I have a bad migration, my local DB is empty and I'd just reset it. I only use the ```prisma migrate dev``` to update my @prisma/client in the front end since I'm using typescript.

Other than that, it's a lot of extra steps, and lost testing data to fix a bad migration. Compare that to running a command in the cli, something like ```"prisma rollback dev <migration name>" //example only in case somebody decides to copy this thinking it's legit```

**Comment by erikhuck** [comment]
+1 for being able to rollback all migrations in a single command similar to the db:migrate:undo:all in the sequelize package. I think prisma is overall better than sequelize but the ability to revert all migrations in a single command is very useful. This is what we'd like to do with our test database so we can use the same test database for multiple apps.

**Reply by erikhuck** [reply]
I've been able to accomplish this by executing `psql -c 'DROP SCHEMA public CASCADE'`. That has the same effect as reverting all migrations because it pretty much deletes all the tables. It would have been nice if prisma simply could revert all the migrations but this solution works well enough.

**Reply by khaelys** [reply]
So basically you would like to have a `migrate reset` that doesn't execute the migrations?

**Comment by abdulhanan001** [comment]
In my case I have delete a pervious migration and then run a command `npx prisma migrate dev` after that I
``` The following migration(s) are applied to the database but missing from the local migrations directory: 20230309223034_add_soft_delete_fields_to_table

? We need to reset the "public" schema at "0.0.0.0:5432"
Do you want to continue? All data will be lost. › (y/N)```

Its really feeling worse to again restore whole  database dump just to rollback or delete one migration :( 

Any update!

**Comment by codePassion-dot** [comment]
it is not just about feeling clever using a library but being able to ship code quickly, hope this would be implemented in the future :disappointed:

**Comment by sagardspeed2** [comment]
Same problem here,
Just revert one column name change, I need to re-dump very large data of the database it not making sense for this small changes.

This rollback feature needs to there for this work.

Eagerly waiting for this feature to introduced in prisma.

**Comment by dukedorje** [comment]
It's been a mainstay in Ruby on Rails for many years, and one of the reasons that ActiveRecord is still the best ORM out there. I'd really like to get full end-to-end Typescript, but until Prisma is up to snuff in its change management, I can't really depend on it for managing large datasets in production.

**Comment by uki1014** [comment]
I often have to modify the migration content, e.g. by reviewing Pull Requests.
I have been using Ruby on Rails for about 4 years now, and in a development environment there are always cases where you have to go back and forth between multiple branches and run different migrations for each one.
The `bin/rails db:rollback` command has been a great tool in those cases.

Is there any reason why Prisma does not (or is difficult to) create a rollback command?

**Comment by suhjohn** [comment]
Seems like https://github.com/sequelize/sequelize supports migrate down, along with any other serious projects like RoR, Django, Go's sql-migrate https://sequelize.org/docs/v6/other-topics/migrations/#undoing-migrations 
Note to anyone who's in the TS ecosystem looking for an actual ORM to use Sequelize. I shouldn't have touched Prisma listening to script kiddies on Reddit.

**Reply by suhjohn** [reply]
For those who use Postgres, I've created the following scripts that effectively automates the annoying parts of the library:

generate-migration.ts: https://gist.github.com/suhjohn/861367eeb0a50cbaba94becaadf4bf09
migrate-down.ts: https://gist.github.com/suhjohn/db1ffba13841dbd775fae83402d645d6

1. Auto-generation of down.sql migration along with the difference from the current database as suggested in this doc (https://www.prisma.io/docs/guides/migrate/developing-with-prisma-migrate/generating-down-migrations) 
2. Execution of down migration and cleanup on _prisma_migrations. 

This assumes the following folder structure:

```
Root
  -- prisma 
    -- migrations
    schema.prisma
  -- scripts
    -- generate-migration.ts
    -- migrate-down.ts
  package.json
```

Hope this is somewhat useful for anyone who find this thread

**Reply by formigarafa** [reply]
That looks nice. I will remember to give it a try if I need to go back to last Prism project I worked with.

**Reply by cevaris** [reply]
Seemingly [Sequelize](https://github.com/sequelize/sequelize/graphs/contributors) main contributors are no longer active.

**Comment by azwar** [comment]
Down migration is easy when we do with other library, so at this time, we use other library to manage db migration. Prisma is used only for query and schema definition. We hope Prisma team has attention about this. Thank you Prisma team.

**Comment by tony-engineering** [comment]
Hello,
I would like to create a generic job that automates database rollbacks. Ideally I would be able to take any point in the code history, and it would rollback the database schema to this state.

Ended up here after reading  
[documentation about down migrations](https://www.prisma.io/docs/guides/migrate/developing-with-prisma-migrate/generating-down-migrations?query=rollback&page=1).
and [production troubleshooting](https://www.prisma.io/docs/guides/migrate/production-troubleshooting)

My conclusion is that Prisma doesn't handle the use case where we want to rollback a migration that **did not** fail. 

Only workaround I can imagine, would be:
1. Force rollback using `migrate db push` from an older state of the code, or create a rollback.sql script with `prisma migrate diff --from-schema-datasource prisma/schema.prisma --to-schema-datamodel prisma/schema.prisma  --script > rollback.sql` then apply it
2. After the "force rollback", if you check status `prisma migrate status` it will always tell you "Database schema is up to date !"
3. To fix that you then need to query the migrations list `SELECT * FROM public._prisma_migrations ORDER BY migration_name ASC`
4. Identify which migrations you rolled back with the rollback.sql script, then delete the corresponding migrations from the table

Alternative to steps 3. and 4. could be `prisma migrate reset`, but this would lead into a data loss. 

At this point Im still unsure how I will implement that.

**Comment by jvolonda42** [comment]
Hello, 

One of the most frustrating thing in this is that the documentation state that we should delete all created migration and create a new one, but this wipe the local database. 

So when stating this:
`
Squashing migrations can be useful when developing with a branch-based workflow. During a large local development effort on a feature branch you might generate multiple migrations using migrate dev. After the feature is finished, the migration history might contain unnecessary intermediate steps that are unwanted in the final migration history that will be pushed to the main branch.
`
For prisma team every complicated feature should be done with a reset of local database?

This needs to be on the top priority.

Thanks for your efforts Prisma team. Your tool is great it just need some missing feature used on every production ORM.

**Comment by foloinfo** [comment]
I wanted something like `rake db:rollback` so I wrote a script. https://gist.github.com/foloinfo/942b8f2c59cb45c355529f68eb63521b
The idea is the same as the [ts script from comment](https://github.com/prisma/prisma/discussions/4617#discussioncomment-6458483), but I'm using basic bash commands and only npx prisma. Now I can freely generate, migrate, rollback, rinse and repeat.

I've never used rollback on production (make a new migration to drop table or change column). So I agree with this comment.

> ...speaking with many developers that most of the time down migrations are not used.
> In a production environment, rollbacks can lead to data loss.

However, I always use rollback in the development environment, maybe I want to tweak the name, made wrong table or something very trivial. Those are not failing migration, but I want to re-do it. It's pretty annoying if I need to reset the development db often.

Maybe this feature is not so important for production use cases, but I think this kind of feature is very important for improving the developer experience, which RoR (ActiveRecord) is very good at.

**Reply by jvolonda42** [reply]
thats what I ended up doing but I really think this should be integrated in prisma and not be done manually by the community.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!

**Comment by colindjk** [comment]
> Hi there,
> 
> To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.
> 
> If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.
> 
> For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.
> 
> Thank you for your understanding and being part of the community!

Could this be reopened? 

I used Prisma ORM at a small company in the past and despite all the benefits Prisma offered with typescript etc, the only strong opinion I had was to get off Prisma ORM before they scaled. Doing database migrations had a negative impact on my blood pressure and anxiety levels. 

If this isn't added, this ORM will just remain a fun to use ORM for typescript fans until they actually have to build something production ready. Or an ORM people mistakenly use until it breaks loudly in a production release. 

The original comment saying devs don't need reverse migration is simply not true, just Google Prisma ORM and see every top result.
